dgraph/cmd/graphql/schema/gqlschema.go

```diff
@@ -275,11 +363,196 @@ func addPatchType(schema *ast.Schema, defn *ast.Definition) {
 	}
 }
 
+// addFieldFilters adds field arguments that allow filtering to all fields of
+// defn that are searchable.  For example, if there's another type
+// `type R { ... f: String @searchable(by: term) ... }`
+// and defn has a field of type R, e.g. if defn is like
+// `type T { ... g: R ... }`
+// then a query should be able to filter on g by term search on f, like
+// query {
+//   getT(id: 0x123) {
+//     ...
+//     g(filter: { f: { anyofterms: "something" } }, first: 10) { ... }
+//     ...
+//   }
+// }
+func addFieldFilters(schema *ast.Schema, defn *ast.Definition) {
+	for _, fld := range defn.Fields {
+		// Filtering makes sense both for lists (= return only items that match
+		// this filter) and for singletons (= only have this value in the result
+		// if it satisfies this filter)
+		addFilterArgument(schema, fld)
+
+		// Ordering and pagination, however, only makes sense for fields of
+		// list types.
+		if fld.Type.Elem != nil {
+			addOrderArgument(schema, fld)
+
+			// Pagination even makes sense when there's no orderables because
+			// Dgraph will do UID order by default.
+			addPaginationArguments(fld)
+		}
+	}
+}
+
+func addFilterArgument(schema *ast.Schema, fld *ast.FieldDefinition) {
+	fldType := fld.Type.Name()
+	if hasSearchables(schema.Types[fldType]) {
+		fld.Arguments = append(fld.Arguments,
+			&ast.ArgumentDefinition{
+				Name: "filter",
+				Type: &ast.Type{NamedType: fldType + "Filter"},
+			})
+	}
+}
+
+func addOrderArgument(schema *ast.Schema, fld *ast.FieldDefinition) {
+	fldType := fld.Type.Name()
+	if hasOrderables(schema.Types[fldType]) {
+		fld.Arguments = append(fld.Arguments,
+			&ast.ArgumentDefinition{
+				Name: "order",
+				Type: &ast.Type{NamedType: fldType + "Order"},
+			})
+	}
+}
+
+func addPaginationArguments(fld *ast.FieldDefinition) {
+	fld.Arguments = append(fld.Arguments,
+		&ast.ArgumentDefinition{Name: "first", Type: &ast.Type{NamedType: "Int"}},
+		&ast.ArgumentDefinition{Name: "offset", Type: &ast.Type{NamedType: "Int"}},
+	)
+}
+
+// addFilterType add a `input TFilter { ... }` type to the schema, if defn
+// is a type that has fields that can be filtered on.  This type filter is used
+// in constructing the corresponding query
+// queryT(filter: TFilter, ... )
+// and in adding search to any fields of this type, like:
+// type R {
+//   f(filter: TFilter, ... ): T
+//   ...
+// }
 func addFilterType(schema *ast.Schema, defn *ast.Definition) {
-	schema.Types[defn.Name+"Filter"] = &ast.Definition{
-		Kind:   ast.InputObject,
-		Name:   defn.Name + "Filter",
-		Fields: getFilterField(),
+	if hasSearchables(defn) {
```

Comment:
Looking at it again, we could have avoided this function because we iterate through all fields here. Do you think the current approach makes for cleaner code?

Reply:
I think if we iterate without checking, then we still have to check this condition at the end before adding to the schema.  Seems better to have it up front.